app/main.py

```diff
@@ -1,214 +1,51 @@
-        "potion": {
-            "name": "Blessing",
-            "effect": {
-                "hp": +10,
-                "power": +5,
-            }
-        }
-    }
-}
-
-
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
-
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
+from app.data.source_data import KNIGHTS
 
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
+from app.knight import Knight
+from app.ammunition.armour import Armour
+from app.ammunition.potion import Potion
+from app.ammunition.weapon import Weapon
 
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
 
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
+def create_knights(knights: dict) -> {Knight}:
+    knights_dict = {}
 
-    # mordred
-    mordred = knightsConfig["mordred"]
+    for params in knights.values():
+        name = params["name"]
+        power = params["power"]
+        hp = params["hp"]
 
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
+        armour = [Armour(armour["part"], armour["protection"])
+                  for armour in params["armour"]]
 
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+        weapon = Weapon(params["weapon"]["name"],
+                        params["weapon"]["power"])
+        try:
+            potion = Potion(params["potion"]["name"],
+                            params["potion"]["effect"])
+        except TypeError:
+            potion = None
 
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
+        knights_dict[name] = Knight(name, power, hp, armour, weapon, potion)
```

Comment:
u can use class attribute here, with all knights, or just delete it from ur class, cause u don't use it anywhere

Reply:
As I understand you talk about ```_all_knights``` class attribute, I use it in class method ```get_hp_all_knights```
If I delete ```knights_dict[name]``` here and will just return class attribute, then in ```battle``` function i'll need to create loop with condition, cause i must specify knights each other by their name
Do I need to do so or somewhat differently ?

Reply:
u didn't fix that